src/lib.rs

```diff
@@ -1,350 +1,434 @@
 use std::collections::HashMap;
+use std::collections::HashSet;
+use std::hash::Hash;
 
-/// Compute the gini impurity of a dataset. 
-/// 
+/// Compute the gini impurity of a dataset.
+///
 /// Returns a float, 0 representing a perfectly pure dataset. Normal distribution: ~0.33
-/// 
+///
 /// By default, any empty dataset will return a gini of 1.0. This may be unexpected behaviour.
 /// ```
 /// use parsnip::gini;
-/// assert_eq!(gini(&vec![0, 0, 0, 1]), 0.375);
+/// assert_eq!(gini(&vec![0_usize, 0, 0, 1]), 0.375);
 /// ```
-pub fn gini(data: &[u64]) -> f32 {
-    if data.len() == 0 {
+pub fn gini<T>(data: &[T]) -> f32
+where
+    T: Eq,
+    T: Hash,
+{
+    if data.is_empty() {
         return 1.0;
-    } 
+    }
     fn p_squared(count: usize, len: f32) -> f32 {
         let p = count as f32 / len;
         p * p
     }
     let len = data.len() as f32;
     let mut count = HashMap::new();
-    for &value in data {
+    for value in data.iter() {
         *count.entry(value).or_insert(0) += 1;
     }
-    let counts: Vec<usize> = count.into_iter().map(|(_, c)| c).collect();
-    let indiv : Vec<f32> = counts.iter().map(|x| p_squared(*x, len)).collect();
-    let sum : f32 = indiv.iter().sum();
+    let sum: f32 = count
+        .into_iter()
+        .map(|(_, c)| c)
+        .map(|x| p_squared(x, len))
+        .sum();
     1.0 - sum
 }
 
 /// The categorical accuracy of a dataset
-/// 
+///
 /// Returns a float where 1.0 is a perfectly accurate dataset
 /// ```
 /// use parsnip::categorical_accuracy;
-/// let pred = vec![0, 0, 0 , 1, 2];
-/// let actual = vec![1, 1, 1, 1, 2];
+/// let pred : Vec<_> = vec![0, 0, 0 , 1, 2];
+/// let actual : Vec<_> = vec![1, 1, 1, 1, 2];
```

Comment:
Since they were built with `vec!`, we already know they are `Vec`s, so these type annotations are redundant. :)

Reply:
done 👍